### ISSUE TITLE
[Parity 12/16] Sparse completeness - CSR/CSC/BSR/BSC, sparse x dense, sparse autograd, 2:4 structured

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet.Tensors#221]
Part of epic #209.

## Summary

`SparseTensor.cs` and `SparseStorageFormat.cs` exist, but the sparse surface is incomplete: we need full CSR/CSC/BSR/BSC support, sparse×dense matmul on both CPU and GPU (cuSPARSE via GPU-primitives issue), sparse autograd, and 2:4 structured sparsity (Ampere+).

Sparse matters for recommendation systems, graph neural nets, scientific computing, sparse fine-tuning (LoRA+sparsity), and inference compression.

### Formats
- COO (✅ we have probably), CSR, CSC, BSR, BSC
- Format conversions (`to_sparse_csr`, `to_sparse_bsr`, etc.)
- Constructors: `sparse_coo_tensor`, `sparse_csr_tensor`, `sparse_csc_tensor`, `sparse_bsr_tensor`, `sparse_bsc_tensor`

### Ops
- `sparse.mm`, `sparse.addmm`, `sparse.bmm`, `sparse.matmul`
- `sparse.sum`, `sparse.softmax`, `sparse.log_softmax`
- `sparse.sampled_addmm` (for attention with sparse mask)
- `sparse.spdiags`
- `sparse_mask` (keep only indices from another sparse tensor)
- `sparse.sum/mean` with `dim`

### Autograd
- Backward through sparse matmul, addmm, softmax — partial in PyTorch, known issues

### 2:4 structured sparsity
- `torch.sparse.SparseSemiStructuredTensor` — Ampere+ hardware `mma.sp` instruction
- Integrated with `torch.compile` for automatic insertion

## Current state in the repo

- `LinearAlgebra/SparseTensor.cs` + `LinearAlgebra/SparseStorageFormat.cs` exist.
- Format coverage, op coverage, autograd coverage all need audit.
- No 2:4 structured sparsity.
- No cuSPARSE binding yet (tracked in GPU-primitives issue).

## How we beat PyTorch

1. **Autotuned sparse-dense tipping point**: `AutotuneCache` measures the crossover between `SparseMM` and `DenseMM` per sparsity pattern + shape + device. PyTorch users must pick manually.
2. **SIMD CSR×dense** with AVX-512 gather+FMA — PyTorch's CPU sparse matmul is notoriously slow. Target 3–5× on CSR×dense.
3. **2:4 structured sparsity on CPU too**: we compile the pattern into an AVX-512 packed representation and beat dense matmul on CPU at 50% sparsity. PyTorch's 2:4 is Ampere-only.
4. **Sparse autograd that compiles**: integrates with `SymbolicBackwardGraph` so sparse→dense→sparse chains simplify.
5. **Sparse attention primitive**: `SampledAddMM` for sparse-masked attention fuses with our FlashAttention-2 kernel family. PyTorch's `sparse.sampled_addmm` doesn't integrate with SDPA.
6. **BSR×dense specialized for block sizes 2/4/8/16** — matches what LLMs actually produce from magnitude-prune workflows. PyTorch's BSR support is basic.
7. **Sparse GGUF / safetensors**: extend the serialization formats to store sparse tensors directly (bonus from serialization issue).
8. **GPU coverage on 7 backends** via cuSPARSE (CUDA), rocSPARSE (HIP), and our own kernels elsewhere.

### Format completeness
- [ ] Audit + complete COO ops
- [ ] CSR / CSC read/write/view
- [ ] BSR / BSC read/write/view (common block sizes: 2, 4, 8, 16)
- [ ] Format conversions (COO ↔ CSR ↔ CSC ↔ BSR ↔ BSC)
- [ ] Constructors surfaced on `Tensor<T>` / `SparseTensor<T>`

### Ops
- [ ] `SparseMatMul(sparse, dense)` — CPU SIMD + CUDA (via cuSPARSE from GPU-primitives issue)
- [ ] `SparseAddMM(input, sparse, dense)`
- [ ] `SparseBmm(sparse_batch, dense_batch)`
- [ ] `SparseSum`, `SparseMean` (dim-aware)
- [ ] `SparseSoftmax`, `SparseLogSoftmax`
- [ ] `SparseMask(tensor, mask_sparse)`
- [ ] `SparseSampledAddMM(alpha * A @ B + beta * C, sparse_pattern)`
- [ ] `SparseSpDiags`
- [ ] `SparseSpMV`, `SparseSpGeMM`

### Autograd
- [ ] Backward rules for all sparse ops
- [ ] Registered through `DifferentiableOps.cs` + `OpRegistry.cs`
- [ ] Gradcheck (dense-jacobian vs sparse-backward)

### 2:4 structured sparsity
- [ ] `SparseSemiStructured<T>` tensor wrapper (50% 2:4 pattern)
- [ ] AVX-512 CPU path (packed representation + matmul)
- [ ] Ampere+ `mma.sp` GPU path
- [ ] Pruning pass that converts a dense weight into 2:4 with minimal accuracy loss (integrate with `BackwardPruningPass` — bonus)

### Autotune
- [ ] Register sparse matmul kernels with `AutotuneCache` for crossover decisions

## Acceptance criteria

- Correctness tests: every format conversion round-trips, every op matches dense reference.
- Gradcheck passes for all differentiable sparse ops.
- CSR×dense benchmark: ≥3× PyTorch CPU sparse at 95% sparsity on 4096×4096.
- 2:4 structured: matches dense result within tolerance; CPU AVX-512 path ≥ 1.3× dense matmul at 50% sparsity.
- FlashAttention-2 + sparse mask integration test (sampled_addmm + our flash kernel).

## Implementation notes

- Extend `LinearAlgebra/SparseTensor.cs`; new format files `SparseCsr.cs`, `SparseCsc.cs`, `SparseBsr.cs`, `SparseBsc.cs` under `LinearAlgebra/Sparse/`.
- SIMD kernels under `Engines/Simd/Sparse/`.
- GPU bindings depend on cuSPARSE / rocSPARSE from the GPU-primitives issue.
- 2:4 structured lives in `LinearAlgebra/Sparse/SemiStructured.cs`.
- Autodiff rules in `Engines/Autodiff/DifferentiableOps.cs`.

**Comment by ooples** [comment]
# Implementation plan — detailed

## 1. Current state

`LinearAlgebra/SparseTensor.cs` + `LinearAlgebra/SparseStorageFormat.cs` exist — coverage needs audit. cuSPARSE binding not present (tracked in #219).

## 2. Scope — full inventory

### Format completeness
```
Audit + complete COO
CSR / CSC read/write/view
BSR / BSC read/write/view (block sizes: 2, 4, 8, 16)
Format conversions (COO ↔ CSR ↔ CSC ↔ BSR ↔ BSC)
Constructors on Tensor<T> / SparseTensor<T>
```

### Ops
```
SparseMatMul(sparse, dense)        # CPU SIMD + CUDA (cuSPARSE via #219)
SparseAddMM(input, sparse, dense)
SparseBmm(sparse_batch, dense_batch)
SparseSum, SparseMean (dim-aware)
SparseSoftmax, SparseLogSoftmax
SparseMask(tensor, mask_sparse)
SparseSampledAddMM(alpha * A @ B + beta * C, sparse_pattern)
SparseSpDiags(diagonals, offsets, shape)
SparseSpMV(sparse, vec)
SparseSpGeMM(A_sparse, B_sparse)
```

### Autograd
```
Backward rules for all sparse ops
Registered through DifferentiableOps.cs + OpRegistry.cs
Gradcheck (dense-jacobian vs sparse-backward)
```

### 2:4 structured sparsity
```
SparseSemiStructured<T> tensor wrapper (50% 2:4 pattern)
AVX-512 CPU path (packed representation + matmul)
Ampere+ `mma.sp` GPU path
Pruning pass: dense → 2:4 minimal accuracy loss (integrate #223 pruning + BackwardPruningPass)
```

### Autotune
```
Register sparse matmul kernels with AutotuneCache for crossover decisions
```

**Total: ~25 ops + format completeness + 2:4 + autotune.**

## 3. Win vectors

1. **Autotuned sparse-dense tipping point** via `AutotuneCache` — measures crossover per (sparsity, shape, device). PyTorch users pick manually.
2. **SIMD CSR×dense** with AVX-512 gather+FMA. Target 3–5× at 95% sparsity 4096×4096.
3. **2:4 structured sparsity on CPU**: compile pattern into AVX-512 packed repr. Beat dense matmul at 50% sparsity. PyTorch 2:4 is Ampere-only.
4. **Sparse autograd that compiles** — `SymbolicBackwardGraph` simplifies sparse→dense→sparse chains.
5. **Sparse attention primitive**: `SampledAddMM` for sparse-masked attention fuses with FlashAttention-2. PyTorch's `sparse.sampled_addmm` doesn't integrate with SDPA.
6. **BSR×dense specialized for block sizes 2/4/8/16** — matches LLM magnitude-prune output. PyTorch BSR is basic.
7. **Sparse GGUF / safetensors**: extend #218 formats to store sparse directly.
8. **GPU coverage on 7 backends** via cuSPARSE + rocSPARSE + our own kernels.

## 4. Per-axis beat targets

| Axis | Target |
|---|---|
| CSR×dense 4096² at 95% sparsity | ≥3× PyTorch CPU sparse |
| SpGEMM 2048² at 99% sparsity | ≥2× PyTorch |
| 2:4 structured CPU at 50% sparsity | ≥1.3× dense matmul |
| 2:4 structured CUDA on Ampere | ≥ parity PyTorch `mma.sp` |
| SampledAddMM + FlashAttention-2 | integrated; PyTorch separate |
| BSR×dense block=4 | ≥2× PyTorch BSR |
| Sparse autograd | gradcheck pass for all ops |
| Autotune sparse↔dense crossover | measurable; PyTorch static |
| All 6 GPU backends | CUDA (cuSPARSE) + HIP (rocSPARSE) + Metal/Vulkan/OpenCL/WebGPU custom kernels; skip-with-reason where needed |

## 5. File inventory

**New (~55):**
- `src/AiDotNet.Tensors/LinearAlgebra/Sparse/` — SparseCsr, SparseCsc, SparseBsr, SparseBsc, FormatConversions, SemiStructured (~10)
- `src/AiDotNet.Tensors/Engines/Simd/Sparse/` — SimdCsrMatMul, SimdBsrMatMul, SimdSpGEMM, Simd24StructMatMul (~8)
- Per-backend GPU sparse kernels (~30)
- `src/AiDotNet.Tensors/LinearAlgebra/Sparse/SamplesAddMM.cs` (for attention)

**Modified (~8):** `SparseTensor.cs`, `SparseStorageFormat.cs`, `OpRegistry.cs`, `DifferentiableOps.cs`, `BackwardFunctions.cs`, `AutotuneCache.cs`, `FusedAttention.cs`.

**Tests (~25):** format conversion round-trip, per-op parity vs dense reference, gradcheck, 2:4 correctness (CPU + CUDA), SampledAddMM + FlashAttention integration.

**Benchmarks (~7):** `CsrMatMulBenchmarks`, `BsrMatMulBenchmarks`, `SpGEMMBenchmarks`, `2_4StructuredBenchmarks`, `SparseAttentionBenchmarks`, `AutotuneCrossoverBenchmarks`, `FormatConversionBenchmarks`.

## 6. Acceptance gate

1. Every format conversion round-trips; every op matches dense reference.
2. Gradcheck passes for all differentiable sparse ops.
3. CSR×dense ≥3× PyTorch at 95% sparsity.
4. 2:4 structured: CPU ≥1.3× dense at 50%, CUDA parity `mma.sp`.
5. SampledAddMM + FlashAttention-2 integration test passes.
6. All 6 GPU backends ship kernels or skip-with-reason.
7. Autotune crossover demonstrable on a sparse Transformer.

## 7. Timeline — ~6 weeks

W1 Format completeness (CSR/CSC/BSR/BSC) + conversions.
W2 Core ops (SparseMatMul/AddMM/Bmm/Sum/Mean).
W3 Softmax/LogSoftmax/Mask/SampledAddMM/SpDiags/SpMV/SpGEMM.
W4 2:4 structured CPU + autotune.
W5 GPU kernels across 6 backends + autograd.
W6 Benchmarks + PR.

## 8. Dependencies

- **Blocks**: #224 (SparseAdam on 2:4), #223 (pruning → 2:4 path).
- **Blocked by**: #219 (cuSPARSE/rocSPARSE bindings), #210 (Sort for top-K in sparse paths), #217 (integration with attention primitives).

## 9. Risks

- **Sparse backward correctness** — PyTorch has known issues; tests must cover density-pattern edge cases.
- **2:4 `mma.sp` GPU** requires Ampere/Hopper — test skip on older HW with documented reason.
- **Autotune crossover overhead** — cache must amortize.

---

*Part of epic #209. Plan ratified 2026-04-19.*
